File: python/src/threads/store/bindings.py
"""Host-issued memory and knowledge bindings (store.sql `memory_bindings`,
`knowledge_bindings`, `provider_audit`).

The host decides which scope owns an item, never the provider: a binding is issued and recorded
here before the provider sees the write, and a returned item counts only when its binding is a
row for the calling scope. Bindings are derived from the scope and the write key, so a retried
write under the same key carries the same binding.
"""

import hashlib
import sqlite3
from typing import Final, Literal

from threads.store.worker import Worker

type Kind = Literal["memory", "knowledge"]

_SQL: Final[dict[Kind, tuple[str, str]]] = {
    "memory": (
        "INSERT OR IGNORE INTO memory_bindings"
        " (namespace, record_id, tenant_id, agent, scope) VALUES (?, ?, ?, ?, ?)",
        "SELECT 1 FROM memory_bindings WHERE namespace = ? AND record_id = ?"
        " AND tenant_id = ? AND agent = ? AND scope = ?",
    ),
    "knowledge": (
        "INSERT OR IGNORE INTO knowledge_bindings"
        " (namespace, record_id, tenant_id, agent, scope) VALUES (?, ?, ?, ?, ?)",
        "SELECT 1 FROM knowledge_bindings WHERE namespace = ? AND record_id = ?"
        " AND tenant_id = ? AND agent = ? AND scope = ?",
    ),
}


def _digest(*parts: str) -> str:
    return hashlib.sha256("\x00".join(parts).encode("utf-8")).hexdigest()[:32]


class Bindings:
    def __init__(self, worker: Worker, kind: Kind) -> None:
        self._worker = worker
        self._kind: Kind = kind
        self._insert, self._owned = _SQL[kind]

    async def issue(self, tenant_id: str, agent: str, scope: str, key: str) -> tuple[str, str]:
        """(namespace, record_id) for a write under `key`, recorded before it is used."""
        namespace = _digest(tenant_id, agent, scope)
        record_id = _digest(namespace, key)
        row = (namespace, record_id, tenant_id, agent, scope)
        await self._worker.call(lambda conn: conn.execute(self._insert, row))
        return namespace, record_id

    async def owned(
        self, tenant_id: str, agent: str, scope: str, items: tuple[tuple[str, str], ...]
    ) -> frozenset[tuple[str, str]]:
        """Which of the (namespace, record_id) pairs belong to the calling scope."""

        def read(conn: sqlite3.Connection) -> frozenset[tuple[str, str]]:
            return frozenset(
                (namespace, record_id)
                for namespace, record_id in items
                if conn.execute(
                    self._owned, (namespace, record_id, tenant_id, agent, scope)
                ).fetchone()
            )

        return await self._worker.call(read)

    async def violation(self, tenant_id: str, namespace: str, record_id: str, at: int) -> None:
        """Records a dropped item in the host's audit table."""
        row = (tenant_id, self._kind, namespace, record_id, at)
        await self._worker.call(
            lambda conn: conn.execute(
                "INSERT INTO provider_audit (tenant_id, kind, code, namespace, record_id, at)"
                " VALUES (?, ?, 'scope_violation', ?, ?, ?)",
                row,
            )
        )

File: typescript/packages/core/src/store/bindings.ts
import { z } from "zod";
import { sha256Hex } from "../hash";
import type { Binding, Scope } from "../memory/protocol";
import { READ_ONLY, type StoreDriver } from "./driver";

// Host-issued bindings (store.sql memory_bindings, knowledge_bindings and // provider_audit). The host, never the provider, decides which records a scope may see: it
// issues a binding before every write or ingest, and before use it looks each returned binding
// up here. A scope label a provider echoes proves nothing, so none is read. Bindings derive from
// the scope and the write key, so a retried write carries the same binding (the same rule, and
// the same bytes, as the Python store: a store is shared by both languages).

export type BindingKind = "memory" | "knowledge";

const TABLE = {
  memory: "memory_bindings",
  knowledge: "knowledge_bindings",
} as const satisfies Record<BindingKind, string>;

const Owned = z.array(z.object({ one: z.literal(1) }));

const digest = (...parts: readonly string[]): string =>
  sha256Hex(parts.join("\u0000")).slice(0, 32);

export class HostBindings {
  readonly #db: StoreDriver;
  readonly #now: () => number;

  constructor(db: StoreDriver, now: () => number) {
    this.#db = db;
    this.#now = now;
  }

  /** The binding of the write under `key` in `scope`, durable before the provider sees it. */
  async issue(kind: BindingKind, scope: Scope, key: string): Promise<Binding> {
    const namespace = digest(scope.tenant_id, scope.agent, scope.scope);
    const binding = { namespace, record_id: digest(namespace, key) };
    await this.#db.transaction((tx) =>
      tx.run(
        `INSERT INTO ${TABLE[kind]} (namespace, record_id, tenant_id, agent, scope)
         VALUES (?, ?, ?, ?, ?) ON CONFLICT DO NOTHING`,
        [
          binding.namespace,
          binding.record_id,
          scope.tenant_id,
          scope.agent,
          scope.scope,
        ],
      ),
    );
    return binding;
  }

  /** Whether the host issued `binding` to exactly this scope. */
  async owns(
    kind: BindingKind,
    scope: Scope,
    binding: Binding,
  ): Promise<boolean> {
    const rows = await this.#db.transaction(
      (tx) =>
        tx.all(
          `SELECT 1 AS one FROM ${TABLE[kind]} WHERE namespace = ? AND record_id = ?
           AND tenant_id = ? AND agent = ? AND scope = ?`,
          [
            binding.namespace,
            binding.record_id,
            scope.tenant_id,
            scope.agent,
            scope.scope,
          ],
        ),
      READ_ONLY,
    );
    return Owned.parse(rows).length > 0;
  }

  /** A hit with an unknown or foreign binding: dropped by the caller, recorded here. */
  async violation(
    kind: BindingKind,
    scope: Scope,
    binding: Binding,
  ): Promise<void> {
    await this.#db.transaction((tx) =>
      tx.run(
        `INSERT INTO provider_audit (tenant_id, kind, code, namespace, record_id, at)
         VALUES (?, ?, 'scope_violation', ?, ?, ?)`,
        [
          scope.tenant_id,
          kind,
          binding.namespace,
          binding.record_id,
          this.#now(),
        ],
      ),
    );
  }
}
